src/async-wrap.cc

```diff
@@ -252,64 +354,63 @@ void LoadAsyncWrapperInfo(Environment* env) {
 
 AsyncWrap::AsyncWrap(Environment* env,
                      Local<Object> object,
-                     ProviderType provider,
-                     AsyncWrap* parent)
-    : BaseObject(env, object), bits_(static_cast<uint32_t>(provider) << 1),
-      uid_(env->get_async_wrap_uid()) {
+                     ProviderType provider)
+    : BaseObject(env, object),
+      provider_type_(provider) {
   CHECK_NE(provider, PROVIDER_NONE);
   CHECK_GE(object->InternalFieldCount(), 1);
 
   // Shift provider value over to prevent id collision.
   persistent().SetWrapperClassId(NODE_ASYNC_ID_OFFSET + provider);
 
-  Local<Function> init_fn = env->async_hooks_init_function();
+  // Use ther Reset() call to call the init() callbacks.
+  Reset();
+}
 
-  // No init callback exists, no reason to go on.
-  if (init_fn.IsEmpty())
-    return;
 
-  // If async wrap callbacks are disabled and no parent was passed that has
-  // run the init callback then return.
-  if (!env->async_wrap_callbacks_enabled() &&
-      (parent == nullptr || !parent->ran_init_callback()))
+AsyncWrap::~AsyncWrap() {
+  if (env()->async_hooks()->fields()[AsyncHooks::kDestroy] == 0) {
     return;
+  }
 
-  HandleScope scope(env->isolate());
-
-  Local<Value> argv[] = {
-    Number::New(env->isolate(), get_uid()),
-    Int32::New(env->isolate(), provider),
-    Null(env->isolate()),
-    Null(env->isolate())
-  };
+  if (env()->destroy_ids_list()->empty())
+    uv_idle_start(env()->destroy_ids_idle_handle(), DestroyIdsCb);
 
-  if (parent != nullptr) {
-    argv[2] = Number::New(env->isolate(), parent->get_uid());
-    argv[3] = parent->object();
-  }
+  env()->destroy_ids_list()->push_back(get_id());
```

Comment:
This shares quite a bit of code with `AddIdToDestroyList`, any chance those could be joined into a common helper?